al_bench/factory.py

```diff
@@ -0,0 +1,138 @@
+# ==========================================================================
+#
+#   Copyright NumFOCUS
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#          https://www.apache.org/licenses/LICENSE-2.0.txt
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ==========================================================================
+
+from __future__ import annotations
+import numpy as np
+import scipy.stats
+from numpy.typing import NDArray
+from typing import Mapping, MutableMapping, Sequence, Any
+
+
+class ComputeCertainty:
+    def __init__(self, certainty_type, percentiles, cutoffs):
+        """
+        certainty_type can be "negative_entropy", "confidence", or "margin", or a list
+        (or tuple) of one or more of these.  A value of None means all certainty types.
+
+        percentiles is a list (or tuple) of percentile values to be computed, e.g., (5,
+        10, 25, 50).  The P percentile score will be that score S such that P percent of
+        all scores are less than (more uncertain than) S.
+
+        cutoffs: if the `certainty_type` is a single type then `cutoffs` can be a list
+        (or tuple) of computed scores to compute the cumulative distribution function
+        value for.  For example for "confidence" it could be (0.5, 0.75, 0.9, 0.95).
+        The cdf value for S will the percentage P of all scores that are less than (more
+        uncertain than) S.  If `certainty_type` includes multiple types (including the
+        case that `certainty_type` is None) then `cutoffs` is a dictionary whose keys
+        are the certainty types and each value is the list (tuple) of cutoffs for that
+        certainty type.
+        """
+
+        all_certainty_types = ["negative_entropy", "confidence", "margin"]
```

Comment:
It might be nice if there was an enum or class or package value that enumerated these rather than having it embedded in the method.  That way, we can display the values from the enum instead of duplicating the values in the consumer code.

Reply:
Once I push it ... I will have made `all_certainty_types` be a static member of the `ComputeCertainty` class, so now it can be easily inspected via something like:
```python
certainty_computer = alb_factory.ComputeCertainty(certainty_type, percentiles, cutoffs)
print(f"all_certainty_types = {certainty_computer.all_certainty_types}")
```
I left the individual certainty types as strings rather than as a (subclass of) Python Enum; I am thinking that is a little easier for the user of this package.  But if you ask again, I'll go ahead and use an Enum.

Reply:
> I left the individual certainty types as strings rather than as a (subclass of) Python Enum; I am thinking that is a little easier for the user of this package. But if you ask again, I'll go ahead and use an Enum.

Strings are fine by me.  I just want to avoid duplicating them (and risking mistyping them).